### PR TITLE
Nds 767 spike react router

### PR DESCRIPTION
**Pull request by coostenbrug**
Intent: Review to merge

This PR:
 - Supports custom render functions in the NavBar
    - API updated to allow for "render", pass in a function
    - custom render is always wrapped with style and onClick
 - Changes spacing styling on Mobile Menu

### CODE REVIEW

File: components/src/NavBar/DesktopMenu.js [hunk 1]
```diff
@@ -6,29 +6,73 @@ import MenuTrigger from "./MenuTrigger";
 import MenuLink from "./MenuLink";
 import theme from "../theme";
 
-const isTrigger = menuItem => (menuItem.items);
+const ApplyMenuLinkStyles = styled.div({
+  "& *": {
```
Comment (matthewlyle) [inline components/src/NavBar/DesktopMenu.js]: i see these in a number of places. does e.g `& *` do anything vs just `*`? i'm pretty sure `& ` renders to nothing. like `li &:hover` would join the parent to be `li:hover`, so `li & *` would join the parent to just be `li *`.

File: components/src/Icon/Icon.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ import PropTypes from "prop-types";
 import styled from "styled-components";
 import { space, color } from "styled-system";
 import icons from "../../icons/icons.json";
+import theme from "../theme";
```
Comment (clemensp) [inline components/src/Icon/Icon.js:6]: I've been wondering about this one:

What's the point of using `ThemeProvider` if we import themes directly from our default theme file?
I think we should be using `ThemeConsumer` to access the theme.

Probably worth opening a ticket for this one.

File: components/src/NavBar/DesktopMenu.js [hunk 3]
```diff
@@ -6,29 +6,73 @@ import MenuTrigger from "./MenuTrigger";
...
+    return (() => (null));
+  }
+};
 
 const BaseDesktopMenu = ({
   menuData,
   ...props
 }) => (
   <Flex { ...props }>
     {menuData.map(menuItem => {
```
Comment (clemensp) [inline components/src/NavBar/DesktopMenu.js]: Could be shortened to:
`menuData.map(menuItem => getRenderFunction(menuItem)(menuItem))`

which could then be shortened to:
`menuData.map(menuItem => renderMenuItem(menuItem))`

which could be shortened to:
`menuData.map(renderMenuItem)`

Thoughts?

File: components/src/NavBar/DesktopMenu.js [hunk 4]
```diff
@@ -6,29 +6,73 @@ import MenuTrigger from "./MenuTrigger";
...
+
+const getRenderFunction = menuItem => {
+  if (menuItem.items) {
+    return renderMenuTrigger;
+  } else if (menuItem.href) {
+    return renderMenuLink;
+  } else if (menuItem.render) {
+    return renderCustom;
+  } else {
+    return (() => (null));
```
Comment (clemensp) [inline components/src/NavBar/DesktopMenu.js:64]: I'm not sure if it's safe to include a `renderNothing` fallback here.
If there's nothing, it's an error, right?

I would consider raising an exception here:
`throw new Error("nothing to render")`?

I'm actually not sure if it's a good practice to throw errors inside render functions, so I would do a quick search for that. I did a 1 minute search, and didn't find much.

File: components/src/NavBar/MobileMenu.js [hunk 5]
```diff
@@ -12,42 +12,128 @@ import MenuLink from "./MenuLink";
...
+  textDecoration: "none",
+  lineHeight: theme.lineHeights.base,
+  transition: ".2s",
+  fontSize: theme.fontSizes.medium,
+  "& *": {
+    display: "block",
+    color: theme.colors.white,
+    textDecoration: "none",
+    padding: `${theme.space.x1} ${theme.space.x2}`,
+    paddingLeft: `${(24 * layer) + 24}px`,
```
Comment (clemensp) [inline components/src/NavBar/MobileMenu.js]: I would extract the padding logic into a function. I see it appear several times.

File: components/src/NavBar/MobileMenu.js [hunk 6]
```diff
@@ -88,48 +179,46 @@ const Menu = styled.ul(() => (
...
+    "&:hover, &:focus": {
+      outline: "none",
+      color: theme.colors.lightBlue,
+      backgroundColor: theme.colors.black,
+      cursor: "pointer",
+    },
+  }
+);
+
 const MobileMenuBase = ({
```
Comment (clemensp) [inline components/src/NavBar/MobileMenu.js]: Nice, looked at the file, and the function reads well.

**Review by clemensp — CHANGES_REQUESTED** [review]
Looks nice.

Some comments left.

Also, make sure you add the changes to `CHANGELOG.md` once you rebase.

File: components/src/NavBar/DesktopMenu.js (hunk 1 above)
Comment (coostenbrug) [reply components/src/NavBar/DesktopMenu.js]: Will make these changes, & will render to nothing while not including & will render to a space with basic nesting (hence why it is needed in the li &:hover example but not in these cases.

File: components/src/Icon/Icon.js (hunk 2 above)
Comment (coostenbrug) [reply components/src/Icon/Icon.js:6]: This makes sense, making a ticket:

https://nulogy-go.atlassian.net/browse/NDS-817

File: components/src/NavBar/DesktopMenu.js (hunk 3 above)
Comment (coostenbrug) [reply components/src/NavBar/DesktopMenu.js]: I cleaned up this instance in DesktopMenu and in MenuTrigger and SubMenuTrigger. The mapping in MobileMenu.js is a bit more complex on how the render function is determined and what is passed to it so I left that one a bit more verbose to keep it readable.

File: components/src/NavBar/DesktopMenu.js (hunk 4 above)
Comment (coostenbrug) [reply components/src/NavBar/DesktopMenu.js:64]: Ideally this is handled with PropTypes. I've made a function to validate that menuData is correct in structure and content and throws errors if not.

**Pull request merged**